### PR TITLE
feat(files): add systemtags view

### LINKED ISSUE #30249: [F2V] Migrate tags view
This ticket is part of the [files to vue migration](https://github.com/nextcloud/server/issues/28063)

### Goal
Unify and improve the system tags view

### QA
- [x] Display tags as folders (new API :warning:)
  - [ ] Implementing this as a DAV might give us opportunity to directly catch DELETE/PUT/MOVE requests
      BUT it will create issues with the paths considering a file located in `dav/USER/files/Photos/vacation.jpg` will be seen with a path like `dav/USER/tags/Summer 2021 vacations/vacation.jpg` while all the actions would expect the real dav path to work :thinking: 
  - [ ] Implementing this with OCS API will allow us to return different data like path that fixes the issue above
- [x] Opening the tag (folder) list all files belonging to it
- [ ] Actions
  - [ ] Rename
  - [ ] Delete
  - [ ] There is no sidebar for those tags/fake-folders, we need to check how to handle this :warning: 
- [x] Assess this global way of managing this. Considering it will use the same API as the real Files/Folders: https://github.com/nextcloud/server/issues/30070, check potential pitfalls of storing and caching "fake folders".
  - [x] Actions might be displayed while not theoretically supporting it on those folders

#### Future
- [ ] Show most recent/used tags in the navigation same as favorites

### LINKED ISSUE #39735: [Bug]: Clicking a folder in the Tags view redirects to home folder view
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Clicking a folder in the Tags view displays its content for a very brief moment and immediately redirects you to the home folder view of the Files app.

### Steps to reproduce

1. Create a folder in the Files app.
2. Tag a folder.
3. Click "Tags" in the left sidebar.
4. Select the tag used in step 2 in the drop-down menu.
5. Click the folder created in step 1.

### Expected behavior

Clicking a folder in the Tags view displays its content. Nothing more should happen.

### Installation method

Community Web installer on a VPS or web space

### Nextcloud Server version

27

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database engine version

MySQL

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "cloud.removed.net",
            "www.cloud.removed.net"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "27.0.1.2",
        "overwrite.cli.url": "https:\/\/cloud.removed.net",
        "overwritehost": "cloud.removed.net",
        "overwriteprotocol": "https",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "proxy": "***REMOVED SENSITIVE VALUE***",
        "proxyexclude": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "connectivity_check_domains": [
            "www.nextcloud.com",
            "www.edri.org"
        ],
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "app_install_overwrite": [
            "apporder",
            "groupfolders",
            "socialsharing_email"
        ],
        "maintenance": false,
        "twofactor_enforced": "false",
        "twofactor_enforced_groups": [],
        "twofactor_enforced_excluded_groups": [],
        "mail_smtpmode": "smtp",
        "mail_sendmailmode": "smtp",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpauth": 1,
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpdebug": true,
        "theme": "",
        "loglevel": 2,
        "logfile": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "tls",
        "updater.release.channel": "stable",
        "default_phone_region": "***REMOVED SENSITIVE VALUE***",
        "trashbin_retention_obligation": "28, 30",
        "data-fingerprint": "11f460aa0538f25a35395a835d034975",
        "versions_retention_obligation": "auto, 180",
        "preview_max_x": 2048,
        "preview_max_y": 2048,
        "preview_max_scale_factor": 1
    }
}
```

### List of activated Apps

```shell
Enabled:
  - activity: 2.19.0
  - admin_audit: 1.17.0
  - apporder: 0.15.0
  - bruteforcesettings: 2.7.0
  - calendar: 4.4.4
  - circles: 27.0.1
  - cloud_federation_api: 1.10.0
  - comments: 1.17.0
  - contacts: 5.3.2
  - contactsinteraction: 1.8.0
  - dav: 1.27.0
  - deck: 1.10.0
  - federatedfilesharing: 1.17.0
  - federation: 1.17.0
  - files: 1.22.0
  - files_automatedtagging: 1.17.0
  - files_linkeditor: 1.1.16
  - files_mindmap: 0.0.28
  - files_pdfviewer: 2.8.0
  - files_rightclick: 1.6.0
  - files_sharing: 1.19.0
  - files_trashbin: 1.17.0
  - files_versions: 1.20.0
  - firstrunwizard: 2.16.0
  - forms: 3.3.1
  - groupfolders: 15.0.1
  - logreader: 2.12.0
  - lookup_server_connector: 1.15.0
  - metadata: 0.19.0
  - nextcloud_announcements: 1.16.0
  - notifications: 2.15.0
  - oauth2: 1.15.1
  - password_policy: 1.17.0
  - polls: 5.2.0
  - privacy: 1.11.0
  - provisioning_api: 1.17.0
  - quota_warning: 1.17.0
  - recommendations: 1.6.0
  - related_resources: 1.2.0
  - richdocuments: 8.1.0
  - serverinfo: 1.17.0
  - settings: 1.9.0
  - sharebymail: 1.17.0
  - socialsharing_email: 2.6.0
  - spreed: 17.0.3
  - systemtags: 1.17.0
  - tables: 0.5.1
  - tasks: 0.15.0
  - text: 3.8.0
  - theming: 2.2.0
  - twofactor_backupcodes: 1.16.0
  - twofactor_totp: 9.0.0
  - updatenotification: 1.17.0
  - user_status: 1.7.0
  - viewer: 2.1.0
  - weather_status: 1.7.0
  - workflowengine: 2.9.0
Disabled:
  - dashboard: 7.7.0 (installed 6.0.0)
  - encryption: 2.15.0
  - extract: 1.3.6 (installed 1.3.6)
  - files_external: 1.19.0
  - files_markdown: 2.4.1 (installed 2.4.1)
  - notes: 4.8.1 (installed 4.8.1)
  - photos: 2.3.0 (installed 1.2.3)
  - ransomware_protection: 1.14.0 (installed 1.14.0)
  - support: 1.10.0 (installed 1.0.0)
  - survey_client: 1.15.0 (installed 1.4.0)
  - suspicious_login: 5.0.0
  - user_ldap: 1.17.0
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
no message logged for this issue
```

### Additional info

Browser console errors
```
XHRPROPFIND
https://cloud.removed.net/remote.php/dav/files/dawa/17
[HTTP/2 404 Not Found 90ms]

XHRGET
https://cloud.removed.net/ocs/v2.php/apps/text/workspace?path=/Talk
[HTTP/2 404 Not Found 96ms]

Content-Security-Policy: The page's settings blocked the loading of a resource at inline ("script-src"). content.js:40:10

GET
https://cloud.removed.net/apps/socialsharing_email/css/socialsharingemail.css?v=783c3f18-34

Content-Security-Policy: The page's settings blocked the loading of a resource at inline ("script-src").
Source: event.preventDefault() 12 files
```

### PR DESCRIPTION
**Pull request by skjnldsv**
Ref: https://github.com/nextcloud/server/issues/39914
- [x] Needs #40474 
- closes https://github.com/nextcloud/server/issues/30249
- closes https://github.com/nextcloud/server/issues/39735

image

### CODE REVIEW

File: apps/files/src/actions/downloadAction.ts [hunk 1]
```diff
@@ -47,7 +47,19 @@ export const action = new FileAction({
...
-		return nodes.length > 0 && nodes
+		if (nodes.length === 0) {
+			return false
+		}
+
+		// We can download direct dav files. But if we have
+		// some folders, we need to use the /apps/files/ajax/download.php
+		// endpoint, which only supports user root folder.
+		if (nodes.some(node => node.type === FileType.Folder)
+			&& !nodes.every(node => node.root?.startsWith('/files'))) {
```
Comment (artonge) [inline apps/files/src/actions/downloadAction.ts:58]: Stops at the first `false`, and is easier to read.

```suggestion
			&& nodes.some(node => !node.root?.startsWith('/files'))) {
```

File: apps/files/src/views/FilesList.vue [hunk 2]
```diff
@@ -373,7 +382,8 @@ export default Vue.extend({
 					this.pathsStore.addPath({ service: currentView.id, fileid: node.fileid, path: join(dir, node.basename) })
 				})
 			} catch (error) {
-				logger.error('Error while fetching content', { error })
+				throw error
+				// logger.error('Error while fetching content', { error })
```
Comment (artonge) [inline apps/files/src/views/FilesList.vue]: Keep or sweep ?

File: apps/systemtags/src/services/systemtags.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		attributes: {
+			...tag,
+			'is-tag': true,
+		},
+	})
+}
+
+export const getContents = async (path = '/'): Promise<ContentsWithRoot> => {
+	// List tags in the root
+	tagsCache = (await fetchTags()).filter(tag => tag.userVisible)
```
Comment (artonge) [inline apps/systemtags/src/services/systemtags.ts:61]: In which case does this `tagsCache` is used as cache ? Seems to me that it is always reset before its usage

File: apps/files/src/views/FilesList.vue (hunk 2 above)
Comment (skjnldsv) [reply apps/files/src/views/FilesList.vue]: Done already :)

File: apps/systemtags/src/services/systemtags.ts (hunk 3 above)
Comment (skjnldsv) [reply apps/systemtags/src/services/systemtags.ts:61]: It's misleading indeed, let me adjust :+1:

**Review by artonge — APPROVED** [review]

**Review by susnux — APPROVED** [review]

**Pull request merged**